File: lib/domain/crop/interactors/masters_nutrient_requirement_show_interactor.rb
# frozen_string_literal: true

module Domain
  module Crop
    module Interactors
      class MastersNutrientRequirementShowInteractor
        def initialize(output_port:, gateway:, requirement_gateway:)
          @output_port = output_port
          @gateway = gateway
          @requirement_gateway = requirement_gateway
        end

        def call(input_dto)
          requirement = @requirement_gateway.find_by_crop_stage_id(input_dto.crop_stage_id)
          if requirement.nil?
            @output_port.on_not_found
          else
            @output_port.on_show_success(requirement)
          end
        end
      end
    end
  end
end

File: crates/agrr-domain/src/crop/interactors/masters_nutrient_requirement_show_interactor.rs
//! Ruby: `Domain::Crop::Interactors::MastersNutrientRequirementShowInteractor`
use crate::crop::dtos::CropStageDetailInput;
use crate::crop::gateways::NutrientRequirementGateway;
use crate::crop::ports::MastersNutrientRequirementOutputPort;

pub struct MastersNutrientRequirementShowInteractor<'a, RG, O> {
    output_port: &'a mut O,
    requirement_gateway: &'a RG,
}

impl<'a, RG, O> MastersNutrientRequirementShowInteractor<'a, RG, O>
where
    RG: NutrientRequirementGateway,
    O: MastersNutrientRequirementOutputPort,
{
    pub fn new(output_port: &'a mut O, requirement_gateway: &'a RG) -> Self {
        Self { output_port, requirement_gateway }
    }

    pub fn call(&mut self, input: CropStageDetailInput) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        match self.requirement_gateway.find_by_crop_stage_id(input.crop_stage_id)? {
            Some(entity) => self.output_port.on_show_success(entity),
            None => self.output_port.on_not_found(),
        }
        Ok(())
    }
}

#[cfg(test)]
mod interactors_masters_nutrient_requirement_show_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/crop/interactors_masters_nutrient_requirement_show_interactor_test.rs"));
}
